gbif-occurrence-linkage/main.py

```diff
@@ -0,0 +1,161 @@
+    }
+    mas_job_record = {
+        "jobId": job_id,
+        "annotations": [annotation]
+    }
+    return mas_job_record
+
+
+def timestamp_now() -> str:
+    """
+    Create a timestamp in the correct format
+    :return: The timestamp as a string
+    """
+    timestamp = str(datetime.now(tz=timezone.utc).strftime("%Y-%m-%dT%H:%M:%S.%f"))
+    timestamp_cleaned = timestamp[:-3]
+    timestamp_timezone = timestamp_cleaned + 'Z'
+    return timestamp_timezone
+
+
+def send_updated_opends(annotation: Dict, producer: KafkaProducer) -> None:
+    """
+    Send the annotation to the Kafka topic
+    :param annotation: The formatted annotationRecord
+    :param producer: The initiated Kafka producer
+    :return: Will not return anything
+    """
+    logging.info('Publishing annotation: ' + str(annotation))
+    producer.send('annotation', annotation)
+
+
+def run_api_call(json_data: Dict) -> Dict[str, str]:
+    """
+    Calls GBIF API based on the occurrenceID, catalogNumber and basisOfRecord
+    :param json_data: The JSON data of the Digital Specimen
+    :return: The result from the API, which contains either the GBIF ID or an error message
+    """
+    identifiers = get_identifiers_from_object(json_data)
+    query_string = (f'https://api.gbif.org/v1/occurrence/search?occurrenceID='
+                    f'{identifiers.get("occurrenceId")}&catalogNumber={identifiers.get("catalogNumber")}'
+                    f'&basisOfRecord={json_data["dwc:basisOfRecord"]}')
+    response = requests.get(query_string)
+    response_json = json.loads(response.content)
+    if response_json['count'] == 1:
+        logging.info('Successfully retrieved a single result from GBIF based on the identifiers')
+        return {'queryString': query_string, 'gbifId': response_json['results'][0]['gbifID']}
+    elif response_json['count'] == 0:
+        logging.info('No results were returned, unable to create a relationship')
+        return {'queryString': query_string, 'error_message': 'Failed to make the match, no match could be created'}
+    else:
+        logging.info('More than one results returned, unable to create a relationship')
+        return {'queryString': query_string, 'error_message': 'Failed to make the match, too many candidates'}
+
+
+def get_identifiers_from_object(json_data: Dict) -> Dict[str, str]:
+    """
+    Retrieve the correct identifiers from the Digital Specimen
+    :param json_data: Json data of the Digital Specimen
+    :return: The mapped identifiers (occurrenceId and catalogNumber)
+    """
+    identifiers = {}
+    for identifier in json_data['identifiers']:
+        if identifier.get('???:identifierType') in ['dwc:occurrenceID', 'abcd:unitGUID']:
+            identifiers['occurrenceId'] = identifier.get('???:identifierValue')
+        if identifier.get('???:identifierType') in ['dwc:catalogNumber', 'abcd:unitID']:
+            identifiers['catalogNumber'] = identifier.get('???:identifierValue')
+    return identifiers
```

Comment:
Is it me or does it seem to overwrite the occurrenceId and catalogNumber props of identifiers on every iteration of the loop?

Reply:
What this does is that it loops over all identifiers. If it finds specific identifiers than it will use those in the request. 
I renamed the variable for the relevant identifiers from `identifiers` to `relevant_identifiers` maybe that helps.

Reply:
That makes sense!